File: Criacionais/Factory Method/Ex001/ex001.py
# Classe Base
class Pessoa:
    def __init__(self, nome):
        self._nome = nome

    def getNome(self):
        return self._nome

    def mostrarPermissao(self):
        raise NotImplementedError("Método deve ser sobrescrito.")


# Subclasse Aluno
class Aluno(Pessoa):
    def __init__(self, nome):
        super().__init__(nome)

    def mostrarPermissao(self):
        print(f"Nome: {self.getNome()}\nCargo: Aluno\nPermissão: Nível 0\n")


# Subclasse Administrativo
class Administrativo(Pessoa):
    def __init__(self, nome):
        super().__init__(nome)

    def mostrarPermissao(self):
        print(f"Nome: {self.getNome()}\nCargo: Administrativo\nPermissão: Nível 1\n")


# Subclasse Professor
class Professor(Pessoa):
    def __init__(self, nome):
        super().__init__(nome)

    def mostrarPermissao(self):
        print(f"Nome: {self.getNome()}\nCargo: Professor\nPermissão: Nível 2\n")


# Subclasse Visitante
class Visitante(Pessoa):
    def __init__(self, nome):
        super().__init__(nome)

    def mostrarPermissao(self):
        print(f"Nome: {self.getNome()}\nCargo: Visitante\nPermissão: Nível 3\n")


# Fábricas
class FabricaDaFatec:
    def criarPessoa(self, nome, cargo):
        cargo = cargo.lower()
        if cargo == "aluno":
            return Aluno(nome)
        elif cargo == "administrativo":
            return Administrativo(nome)
        elif cargo == "professor":
            return Professor(nome)
        elif cargo == "visitante":
            return Visitante(nome)
        else:
            raise ValueError("Cargo desconhecido.")


# USO DA FABRICA - CLIENTE:
class Main:
    @staticmethod
    def main():
        fabrica = FabricaDaFatec()

        p1 = fabrica.criarPessoa("Carlos", "aluno")
        p2 = fabrica.criarPessoa("Ana", "aluno")
        p3 = fabrica.criarPessoa("Mariana", "administrativo")
        p4 = fabrica.criarPessoa("João", "administrativo")
        p5 = fabrica.criarPessoa("Roberto", "professor")
        p6 = fabrica.criarPessoa("Patrícia", "professor")
        p7 = fabrica.criarPessoa("Lucas", "visitante")
        p8 = fabrica.criarPessoa("Fernanda", "visitante")

        p1.mostrarPermissao()
        p2.mostrarPermissao()
        p3.mostrarPermissao()
        p4.mostrarPermissao()
        p5.mostrarPermissao()
        p6.mostrarPermissao()
        p7.mostrarPermissao()
        p8.mostrarPermissao()


if __name__ == "__main__":
    Main.main()

File: Criacionais/Factory Method/Ex001/ex001.js
// Classe Base
class Pessoa {
    constructor(nome) {
        this.nome = nome;
    }

    getNome() {
        return this.nome;
    }

    mostrarPermissao() {
        throw new Error("Método deve ser sobrescrito.");
    }
}

// Subclasse Aluno
class Aluno extends Pessoa {
    constructor(nome) {
        super(nome);
    }

    mostrarPermissao() {
        console.log(`Nome: ${this.getNome()}\nCargo: Aluno\nPermissão: Nível 0\n`);
    }
}

// Subclasse Administrativo
class Administrativo extends Pessoa {
    constructor(nome) {
        super(nome);
    }

    mostrarPermissao() {
        console.log(`Nome: ${this.getNome()}\nCargo: Administrativo\nPermissão: Nível 1\n`);
    }
}

// Subclasse Professor
class Professor extends Pessoa {
    constructor(nome) {
        super(nome);
    }

    mostrarPermissao() {
        console.log(`Nome: ${this.getNome()}\nCargo: Professor\nPermissão: Nível 2\n`);
    }
}

// Subclasse Visitante
class Visitante extends Pessoa {
    constructor(nome) {
        super(nome);
    }

    mostrarPermissao() {
        console.log(`Nome: ${this.getNome()}\nCargo: Visitante\nPermissão: Nível 3\n`);
    }
}

// Fábricas
class FabricaDaFatec {
    criarPessoa(nome, cargo) {
        switch (cargo.toLowerCase()) {
            case "aluno":
                return new Aluno(nome);
            case "administrativo":
                return new Administrativo(nome);
            case "professor":
                return new Professor(nome);
            case "visitante":
                return new Visitante(nome);
            default:
                throw new Error("Cargo desconhecido.");
        }
    }
}

// USO DA FABRICA - CLIENTE:
class Main {
    static main() {
        const fabrica = new FabricaDaFatec();

        const p1 = fabrica.criarPessoa("Carlos", "aluno");
        const p2 = fabrica.criarPessoa("Ana", "aluno");
        const p3 = fabrica.criarPessoa("Mariana", "administrativo");
        const p4 = fabrica.criarPessoa("João", "administrativo");
        const p5 = fabrica.criarPessoa("Roberto", "professor");
        const p6 = fabrica.criarPessoa("Patrícia", "professor");
        const p7 = fabrica.criarPessoa("Lucas", "visitante");
        const p8 = fabrica.criarPessoa("Fernanda", "visitante");

        p1.mostrarPermissao();
        p2.mostrarPermissao();
        p3.mostrarPermissao();
        p4.mostrarPermissao();
        p5.mostrarPermissao();
        p6.mostrarPermissao();
        p7.mostrarPermissao();
        p8.mostrarPermissao();
    }
}

Main.main();